### ISSUE TITLE
Do we need sub-populations?

### ISSUE THREAD

**Issue by leonlan** [PyVRP/PyVRP#284]
Feasible and infeasible solutions are currently divided into separate sub-populations. But do we really need this? A very old version of PyVRP did not contain sub-populations, but instead used one single population that contained all solutions. We then added sub-populations again (https://github.com/N-Wouda/Euro-NeurIPS-2022/pull/72) citing performance reasons, although I recall that these improvements were very small (in the range of 0.05-0.10% for <=15 min time limits).

I'll run an experiment to see what the marginal contribution is of having a sub-population. This can be done easily by adding individuals to the feasible sub-population, by default. If the marginal contribution is low, we might want to consider removing the sub-population altogether.

**Comment by N-Wouda** [comment]
See also #87 for related stuff.

**Comment by leonlan** [comment]
I'm running the following three configurations to see what's the impact of removing subpopulations:
- default config.
- min pop size 50, generation size 80.
- min pop size 50, generation size 80, nb_elite 8.

I simply changed `Population.add` to always add individuals to the "feasible" population. I'll know more in a few hours!

**Comment by N-Wouda** [comment]
Did you also update `Population._get_tournament` to never look at the infeas solutions?

**Comment by leonlan** [comment]
GH x 10 seeds x 1 hour
|       | main  | no subpop | subpop + larger pop size | subpop + larger pop size + more elite |
|-------|-------|-----------|--------------------------|---------------------------------------|
| gap   | 0.52% | 0.62%     | 0.66%                    | 0.60%                                 |
| iters | 57015 | 59009     | 46833                    | 58107                                 |


Separating feasible and infeasible solutions into separate subpopulations is indeed better, with a ~0.10% gap difference, as also observed during our EURO-NeurIPS submission. But this may also be due to the repair/penalty management that's fitted on the old situation. I'll have to think a bit about how to adjust those parameters and then I'll run a few more experiments.

**Comment by leonlan** [comment]
> Did you also update `Population._get_tournament` to never look at the infeas solutions? Since those never get updated, they would be really bad in crossover.

I'm not sure what you mean by this. I simply add both feasible and infeasible solutions to the feasible subpopulation, effectively having just one population. The infeasible solutions are still used for crossover.

**Comment by N-Wouda** [comment]
Nvm, I wasn't sure how we added infeasible initial solutions in GA. I just checked and that comment doesn't make any sense.

**Comment by N-Wouda** [comment]
There's definitely some interaction between population size and the number of iterations between penalty updates. I'm not sure what else there might be.

**Comment by leonlan** [comment]
This is a bit more tricky than I thought it would be. My hypothesis is that having feasible/infeasible solutions in the same population gives an advantage to the infeasible solutions to survive, because their cost is generally lower due to the allowed infeasibilities.

I ran a few more experiments to get a feeling of which parameters influence the whole process, but it's not been successful. The default config here was:
- min pop size 50, generation size 80, nb_elite 8: 0.60%

Then I tried the following:
- 50->100 iters for penalty updates: 0.63%
- 40/60 population size: 0.63% 
- 100 iters penalty and 40/60 population size: 0.66%
- nbClose 5->10: 0.63%
- targetFeasible 0.43 -> 0.6: 0.58%
- decreasePenalty: 0.34 -> 0.75: 0.62%

Without a solution to infeasible solutions being favored over feasible due to cost, I don't see this working out. Having separate subpopulations is a way to do this without introducing any penalty parameters. So I think the conclusion here is to just keep the subpopulations.
